### PR TITLE
Fixes issue  #31

### PR DESCRIPTION
**Pull request by MrM0nkey**
This PR should fix the bug mentioned in  #31.

### CODE REVIEW

**Comment by benj-zen** [comment]
Isn't the bias already tested in test_simulation?

**Comment by MrM0nkey** [comment]
@benj-zen @nils-wisiol There is a test for [LTFArray](https://github.com/nils-wisiol/pypuf/blob/fix_bias_bug/test/test_simulation.py#L454-L488) but not for SimulationMajorityLTFArray. I do not like the test for LTFArray because it just mimics the function. We should check the behavior of the output instead of just repelicate it. I have to reflect over a better test.

File: pypuf/simulation/arbiter_based/ltfarray.py [hunk 1]
```diff
@@ -500,15 +500,22 @@ def __init__(self, weight_array, transform, combiner, bias=False):
...
         :param inputs: list of challenges
         :return: list of responses
         """
-        if self.bias:
-            inputs = tools.iter_append_last(inputs, 1)
         return sign(self.val(inputs))
 
     def val(self, inputs):
+        """
+        This function a calculates the output of the LTFArray based on weights.
```
Comment (nils-wisiol) [inline pypuf/simulation/arbiter_based/ltfarray.py]: 'a': typo

**Comment by MrM0nkey** [comment]
```bias = normal_bias(mu, var)
LTFArray(....,bias=bias)
```
or also

```LTFArray(....,bias=3.7)```

was my first proposal....

**Comment by benj-zen** [comment]
As I said... Misunderstanding :/
@nils-wisiol what is your opinion on that?

**Comment by MrM0nkey** [comment]
@benj-zen please review this PR again.

File: pypuf/experiments/experiment/majority_vote.py [hunk 2]
```diff
@@ -60,8 +60,9 @@ def __init__(self, log_name, n, k, challenge_count, seed_instance, seed_instance
         :param iterations: int
                            The number of evaluations of the SimulationMajorityLTFArray instance which are used
                            to check the desired_stability.
-        :param bias: boolean
-                     The value is used to turn on input/output distort of the  PUF instance simulation.
+        :param bias: None, float or an one dimensional list of float with shape (k)
```
Comment (benj-zen) [inline pypuf/experiments/experiment/majority_vote.py]: **a** one dimensional

Comment (benj-zen) [inline pypuf/experiments/experiment/majority_vote.py]: **a** one dimensional

File: pypuf/simulation/arbiter_based/ltfarray.py [hunk 3]
```diff
@@ -622,28 +622,38 @@ def normal_weights(n, k, mu=0, sigma=1, random_instance=RandomState()):
...
 
-    def __init__(self, weight_array, transform, combiner, bias=False):
+    def __init__(self, weight_array, transform, combiner, bias=None):
         """
         Initializes an LTFArray based on given weight_array and
         combiner function with appropriate transformation of challenges.
         The bias is committed through the (n+1)th value in weight_array.
         So the parameter bias only states if the given weight_array has
         n+1 values (or not) while the challenges still has length n.
+        :param bias: None, float or an two dimensional array of float with shape (k, 1)
```
Comment (benj-zen) [inline pypuf/simulation/arbiter_based/ltfarray.py]: **a** two dimensional

File: pypuf/simulation/arbiter_based/ltfarray.py [hunk 4]
```diff
@@ -685,11 +699,14 @@ def sigma_noise_from_random_weights(n, sigma_weight, noisiness=0.1):
         return sqrt(n) * sigma_weight * noisiness
 
     def __init__(self, weight_array, transform, combiner, sigma_noise,
-                 random_instance=RandomState(), bias=False):
+                 random_instance=RandomState(), bias=None):
         """
         Initializes LTF array like in LTFArray and uses the provided
         PRNG instance for drawing noise values. If no PRNG provided, a
         fresh `numpy.random.RandomState` instance is used.
+        :param bias: None, float or an two dimensional array of float with shape (k, 1)
```
Comment (benj-zen) [inline pypuf/simulation/arbiter_based/ltfarray.py]: **a** two dimensional

File: pypuf/simulation/arbiter_based/ltfarray.py [hunk 5]
```diff
@@ -732,8 +749,9 @@ def __init__(self, weight_array, transform, combiner, sigma_noise,
                             Standard deviation of noise distribution.
         :param random_instance_noise: RandomState
                                       This pseudo-random number generator is used to generate noise.
-        :param bias: boolean
-                     This value is used to turn on input/output distort of this simulation.
+        :param bias: None, float or an two dimensional array of float with shape (k, 1)
```
Comment (benj-zen) [inline pypuf/simulation/arbiter_based/ltfarray.py]: **a** two dimensional

File: test/test_experiment.py [hunk 6]
```diff
@@ -87,11 +87,96 @@ def test_run_and_analyze(self):
...
 
         self.assertGreaterEqual(experiment.result_overall_stab, experiment.overall_desired_stability,
                                 'No vote_count was found.')
+
+        queue.put_nowait(None)
+        listener.join()
+
+    def test_run_and_analyze_bias_list(self):
+        """
+        This method run the experiment with a bias list and checks if a number of votes was found in order to satisfy an
```
Comment (benj-zen) [inline test/test_experiment.py]: **runs** the experiment

File: test/test_experiment.py [hunk 7]
```diff
@@ -87,11 +87,96 @@ def test_run_and_analyze(self):
...
+
+        self.assertGreaterEqual(experiment.result_overall_stab, experiment.overall_desired_stability,
+                                'No vote_count was found.')
+
+        queue.put_nowait(None)
+        listener.join()
+
+    def test_run_and_analyze_bias_value(self):
+        """
+        This method run the experiment with a bias value and checks if a number of votes was found in order to
```
Comment (benj-zen) [inline test/test_experiment.py]: **runs** the experiment

File: test/test_simulation.py [hunk 8]
```diff
@@ -471,49 +472,94 @@ class TestLTFArray(unittest.TestCase):
...
+        (64, 4, 0, 1, 1.5),
+        (4, 16, 0, 1, 2.3),
+        (16, 4, 10, .5, array([1.4, 1.0, 0.5, 0.4])),
     ]
 
-    def test_bias(self):
+    def test_bias_influence_value(self):
         """
-        Probabilistic test for checking the bias feature in eval.
+        This method test the influence of the bias value. The results should be different.
```
Comment (benj-zen) [inline test/test_simulation.py]: This method **tests**

File: test/test_simulation.py [hunk 9]
```diff
@@ -471,49 +472,94 @@ class TestLTFArray(unittest.TestCase):
...
+        sigma = 0.5
 
-        for test_parameters in self.test_set:
-            n = test_parameters[0]
-            k = test_parameters[1]
-            mu = test_parameters[2]
-            sigma = test_parameters[3]
-            bias = test_parameters[4]
-            weight_array = LTFArray.normal_weights(n, k, mu, sigma)
+        challanges = array(list(tools.all_inputs(n)))
```
Comment (benj-zen) [inline test/test_simulation.py]: **challenges**

File: test/test_simulation.py [hunk 10]
```diff
@@ -471,49 +472,94 @@ class TestLTFArray(unittest.TestCase):
...
+    def test_bias_influence_array(self):
+        """
+        This method test the influence of the bias array. The results should be different.
+        """
+        n = 8
+        k = 4
+        mu = 1
+        sigma = 0.5
+
+        challanges = array(list(tools.all_inputs(n)))
```
Comment (benj-zen) [inline test/test_simulation.py]: **challenges**

File: test/test_simulation.py [hunk 11]
```diff
@@ -614,6 +660,91 @@ def test_ltf_eval(self):
                 around(noisy_ltf_array.ltf_eval(transformed_inputs), decimals=10)
             )
 
+    def test_bias_influence_array(self):
+        """
+        This method test the influence of the bias array. The results should be different.
```
Comment (benj-zen) [inline test/test_simulation.py]: This method **tests**

File: test/test_simulation.py [hunk 12]
```diff
@@ -614,6 +660,91 @@ def test_ltf_eval(self):
...
+        """
+        n = 8
+        k = 4
+        mu = 1
+        sigma = 0.5
+
+        challanges = array(list(tools.all_inputs(n)))
```
Comment (benj-zen) [inline test/test_simulation.py]: **challenges**

File: test/test_simulation.py [hunk 13]
```diff
@@ -614,6 +660,91 @@ def test_ltf_eval(self):
...
+
+        biased_responses = biased_ltf_array.eval(challanges)
+        responses = ltf_array.eval(challanges)
+
+        # The arithmetic mean of the res
+        self.assertFalse(array_equal(biased_responses, responses))
+
+    def test_bias_influence_value(self):
+        """
+        This method test the influence of the bias value. The results should be different.
```
Comment (benj-zen) [inline test/test_simulation.py]: This method **tests**

File: test/test_simulation.py [hunk 14]
```diff
@@ -614,6 +660,91 @@ def test_ltf_eval(self):
...
+        """
+        n = 8
+        k = 4
+        mu = 1
+        sigma = 0.5
+
+        challanges = array(list(tools.all_inputs(n)))
```
Comment (benj-zen) [inline test/test_simulation.py]: **challenges**

File: test/test_simulation.py [hunk 15]
```diff
@@ -712,3 +843,95 @@ def get_functions_with_prefix(prefix, obj):
                     vote_count=vote_count,
                 )
                 mv_noisy_ltf_array.eval(inputs)
+
+    def test_bias_influence(self):
+        """
+        This method test the influence of the bias array. The results should be different.
```
Comment (benj-zen) [inline test/test_simulation.py]: This method **tests**

File: test/test_simulation.py [hunk 16]
```diff
@@ -712,3 +843,95 @@ def get_functions_with_prefix(prefix, obj):
...
+        """
+        n = 8
+        k = 4
+        mu = 1
+        sigma = 0.5
+
+        challanges = challanges = array(list(tools.all_inputs(n)))
```
Comment (benj-zen) [inline test/test_simulation.py]: **challenges**

File: test/test_simulation.py [hunk 17]
```diff
@@ -712,3 +843,95 @@ def get_functions_with_prefix(prefix, obj):
...
+
+        biased_responses = biased_ltf_array.eval(challanges)
+        responses = ltf_array.eval(challanges)
+
+        # The arithmetic mean of the res
+        self.assertFalse(array_equal(biased_responses, responses))
+
+    def test_bias_influence_value(self):
+        """
+        This method test the influence of the bias value. The results should be different.
```
Comment (benj-zen) [inline test/test_simulation.py]: This method **tests**

File: test/test_simulation.py [hunk 18]
```diff
@@ -712,3 +843,95 @@ def get_functions_with_prefix(prefix, obj):
...
+        """
+        n = 8
+        k = 4
+        mu = 1
+        sigma = 0.5
+
+        challanges = array(list(tools.all_inputs(n)))
```
Comment (benj-zen) [inline test/test_simulation.py]: **challenges**

File: test/test_tools.py [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+        arr_res = append_last(arr, item)
+        self.assertEqual(arr_res.dtype, typ, 'The types must be equal.')
+
+        arr = zeros((3, 2, 1), dtype=typ)
+        item = 30.1
+        with self.assertRaisesRegex(AssertionError, 'The elements of arr and item must be of the same type.'):
+            append_last(arr, item)
+
+    def test_proper_dimension(self):
+        """This function checks if the function recognize that the array have no dimensions."""
```
Comment (benj-zen) [inline test/test_tools.py]: the function **recognizes** that the array **has dimension zero.**

**Comment by MrM0nkey** [comment]
@benj-zen I have fixed the typos.

File: test/test_simulation.py [hunk 20]
```diff
@@ -471,49 +472,94 @@ class TestLTFArray(unittest.TestCase):
...
+            combiner=LTFArray.combiner_xor,
+            bias=None,
+        )
+        # the second dimension of the weight_array plus one must be the number of elements in biased weight_array
+        self.assertEqual(shape(ltf_array.weight_array)[1]+1, shape(biased_ltf_array.weight_array)[1])
+
+        bias_array = biased_ltf_array.weight_array[:, -1]
+        bias_array_compared = [bias == bias_array[0] for bias in bias_array]
+        # the bias values should be equal for this test.
+        self.assertTrue(reduce((lambda bool_1, bool_2: bool_1 and bool_2), bias_array_compared))
```
Comment (nils-wisiol) [inline test/test_simulation.py]: easier: `numpy.array(bias_array_compared).all()`

File: test/test_simulation.py [hunk 21]
```diff
@@ -614,6 +660,91 @@ def test_ltf_eval(self):
...
+            sigma_noise=sigma,
+            bias=None,
+        )
+        # the second dimension of the weight_array plus one must be the number of elements in biased weight_array
+        self.assertEqual(shape(ltf_array.weight_array)[1]+1, shape(biased_ltf_array.weight_array)[1])
+
+        bias_array = biased_ltf_array.weight_array[:, -1]
+        bias_array_compared = [bias == bias_array[0] for bias in bias_array]
+        # the bias values should be equal for this test.
+        self.assertTrue(reduce((lambda bool_1, bool_2: bool_1 and bool_2), bias_array_compared))
```
Comment (nils-wisiol) [inline test/test_simulation.py]: same

File: test/test_simulation.py [hunk 22]
```diff
@@ -712,3 +843,95 @@ def get_functions_with_prefix(prefix, obj):
...
+            bias=None,
+        )
+        # the second dimension of the weight_array plus one must be the number of elements in biased weight_array
+        self.assertEqual(shape(ltf_array.weight_array)[1] + 1, shape(biased_ltf_array.weight_array)[1])
+
+        bias_array = biased_ltf_array.weight_array[:, -1]
+        bias_array_compared = [bias == bias_array[0] for bias in bias_array[1:]]
+        # the bias values should be different for this test. It is possible that they are all equal but this chance is
+        # low.
+        self.assertFalse(reduce((lambda bool_1, bool_2: bool_1 and bool_2), bias_array_compared))
```
Comment (nils-wisiol) [inline test/test_simulation.py]: same

File: test/test_simulation.py [hunk 23]
```diff
@@ -712,3 +843,95 @@ def get_functions_with_prefix(prefix, obj):
...
+            random_instance_noise=RandomState(0xCCABAD),
+            bias=None,
+        )
+        # the second dimension of the weight_array plus one must be the number of elements in biased weight_array
+        self.assertEqual(shape(ltf_array.weight_array)[1] + 1, shape(biased_ltf_array.weight_array)[1])
+
+        bias_array = biased_ltf_array.weight_array[:, -1]
+        bias_array_compared = [bias == bias_array[0] for bias in bias_array]
+        # the bias values should be equal for this tests.
+        self.assertTrue(reduce((lambda bool_1, bool_2: bool_1 and bool_2), bias_array_compared))
```
Comment (nils-wisiol) [inline test/test_simulation.py]: same

**Pull request merged**